perf/benches/recycler.rs

```diff
@@ -0,0 +1,23 @@
+#![feature(test)]
+
+extern crate test;
+
+use solana_perf::recycler::Recycler;
+use test::Bencher;
+
+#[bench]
+fn bench_recycler(bencher: &mut Bencher) {
+    solana_logger::setup();
+
+    let recycler = Recycler::default();
+
+    for _ in 0..1000 {
+        let mut p = solana_perf::packet::Packets::with_capacity(128);
+        p.packets.resize(128, solana_sdk::packet::Packet::default());
+        recycler.recycle_for_test(p.packets);
+    }
+
+    bencher.iter(move || {
+        recycler.recycle_for_test(recycler.allocate("me"));
```

Comment:
> Nice, this seems pretty good, any perf change with the vec-deque and extra checks?

In summary, sadly this increased twice of fast path of recycle->allocate iteration around 70ns to 140ns. But, I think this shouldn't affect the overall cluster perf. (I'm running gcp system-performance-test to be sure).

In detail, simply changing to VecDeque increases around 10ns and 30ns * 2 by 2 Instance::now() calls by recycle and allocate.

```
base:
test bench_recycler ... bench:          73 ns/iter (+/- 22)

only changed Vec => VecDeque collection type:
test bench_recycler ... bench:          85 ns/iter (+/- 7)

this pr:
test bench_recycler ... bench:         144 ns/iter (+/- 19)
```

I think these numbers are reasonable. VecDec is basically Vec and it only some small additional offset calculation so the order should be 10ns-20ns.

As for Instance::now, I think this is decent value as well, considering it's basically vdso call, recalling my vague perf number for it. xD

Also, I also think `VecDeque` is faster than `BTreeMap` or `BinaryHeap` implementation. Although, I haven't tested.

This tested on my laptop and I know there is jitter. If you're curious, I can run more rigorous experiment in more clean room enviroment. :)

Reply:
ok not bad, thanks.